deploy/compile-zk.ts

```diff
@@ -0,0 +1,106 @@
+import { utils, Wallet, Provider, EIP712Signer, types } from "zksync-web3";
+import * as ethers from "ethers";
+import { HardhatRuntimeEnvironment } from "hardhat/types";
+import { CONTRACT_DEPLOYER_ADDRESS } from "zksync-web3/build/src/utils";
+import { Deployer } from "@matterlabs/hardhat-zksync-deploy";
+import dotenv from "dotenv";
+
+const NONCE_HOLDER_SYSTEM_CONTRACT = "0x0000000000000000000000000000000000008003";
+const NONCE_HOLDER_SYSTEM_CONTRACT_ABI = [ "function getDeploymentNonce(address _address) external view returns (uint256 deploymentNonce)" ];
+
+dotenv.config();
+
+export default async function signDeployFactoryContractTX(hre: HardhatRuntimeEnvironment) {
+  const RPC = process.env.RPC;
+  const MNEMONIC = process.env.MNEMONIC;
+  const PK = process.env.PK;
+
+  if (!RPC) {
+    throw new Error('Missing RPC environment variable');
+  }
+  const provider = new Provider(RPC);
+  let wallet;
+  //wallet = Wallet.createRandom();
+  if (PK) {
+    wallet = new Wallet(PK).connect(provider);
+  } else if (MNEMONIC) {
+    wallet =  Wallet.fromMnemonic(MNEMONIC).connect(provider);
+  } else {
+    throw new Error('Either PK or MNEMONIC environment variable must be set');
+  }
+
+  const fromAddress = wallet.address;
+
+  const deployer = new Deployer(hre, wallet);
+  const factoryArtifact = await deployer.loadArtifact("SafeSingeltonFactory");
+
+  const salt = ethers.constants.HashZero;
+  const bytecodeHash = utils.hashBytecode(factoryArtifact.bytecode);
+  const input = "0x";
+  // We use create2 here as the address of this will be zkSync specific in any case. This way it also provides additional security.
+  const functionSelector = ethers.utils.solidityKeccak256(['string'], ['create2(bytes32,bytes32,bytes)']).slice(0, 10);
+  const encodedData = new ethers.utils.AbiCoder().encode(
+    ['bytes32', 'bytes32', 'bytes'],
+    [salt, bytecodeHash, input]
+  ).slice(2);
+  const data = functionSelector + encodedData;
```

Comment:
this works but imo it's more elegant to use the abicoder directly to encode the calldata

Reply:
Fixed by https://github.com/protofire/safe-singleton-factory/commit/21fb0a2f6e0deb787712fd2e7ace8fce2906f800